File: C++/95. Unique Binary Search Trees II.cpp
// Given an integer n, return all the structurally unique BST's (binary search trees), which has exactly n nodes of unique values from 1 to n. Return the answer in any order.

 

// Example 1:


// Input: n = 3
// Output: [[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]
// Example 2:

// Input: n = 1
// Output: [[1]]
 

// Constraints:

// 1 <= n <= 8


#include <iostream>
#include <vector>
using namespace std;



// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int val = 0, TreeNode* left = nullptr, TreeNode* right = nullptr)
        : val(val), left(left), right(right) {}
};

std::vector<TreeNode*> generateTrees(int n) {
    if (n == 0) {
        return std::vector<TreeNode*>();
    }

    std::function<std::vector<TreeNode*>(int, int)> dfs = [&](int start, int end) {
        std::vector<TreeNode*> ans;

        if (start > end) {
            ans.push_back(nullptr);
            return ans;
        }

        for (int i = start; i <= end; ++i) {
            std::vector<TreeNode*> leftTree = dfs(start, i - 1);
            std::vector<TreeNode*> rightTree = dfs(i + 1, end);

            for (auto& left : leftTree) {
                for (auto& right : rightTree) {
                    TreeNode* root = new TreeNode(i);
                    root->left = left;
                    root->right = right;
                    ans.push_back(root);
                }
            }
        }

        return ans;
    };

    return dfs(1, n);
}

File: Javascript/95. Unique Binary Search Trees II.js
// Given an integer n, return all the structurally unique BST's (binary search trees), which has exactly n nodes of unique values from 1 to n. Return the answer in any order.

 

// Example 1:


// Input: n = 3
// Output: [[1,null,2,null,3],[1,null,3,2],[2,1,3],[3,1,null,null,2],[3,2,null,1]]
// Example 2:

// Input: n = 1
// Output: [[1]]


/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var generateTrees = function(n) {
    function dfs(nums) {
      if (!nums.length) {
        return [null];
      }
      const ans = [];
      for (let i = 0; i < nums.length; i++) {
        const leftTree = dfs(nums.slice(0, i));
        const rightTree = dfs(nums.slice(i + 1));
        for (const l of leftTree) {
          for (const r of rightTree) {
            const root = new TreeNode(nums[i]);
            root.left = l;
            root.right = r;
            ans.push(root);
          }
        }
      }
      return ans;
    }

    const nums = Array.from({ length: n }, (_, i) => i + 1);
    return dfs(nums);
};

